### ISSUE TITLE
function :options.exception/1 is undefined (module :options is not available)

### ISSUE THREAD

**Issue by fbettag** [kipcole9/money#137]
Hi there,

i am encountering this issue only production. on my local development machine, everything works as expected.

```
** (exit) an exception was raised:
    ** (UndefinedFunctionError) function :options.exception/1 is undefined (module :options is not available)
        :options.exception({:socket_options, [{:log_level, :alert}, :inet, {:packet_size, 0}, {:packet, 0}, {:header, 0}, {:active, false}, {:mode, :binary}]})
        (ex_money 5.7.0) lib/money.ex:1745: Money.to_currency!/3
```

Production is OpenBSD 7.0 (amd64):
```
Erlang/OTP 21 [erts-10.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:1]

Elixir 1.11.4 (compiled with Erlang/OTP 21)
```

Development is a current ArchLinux (amd64):
```
Erlang/OTP 24 [erts-12.1.5] [source] [64-bit] [smp:12:12] [ds:12:12:10] [async-threads:1] [jit]

Elixir 1.12.3 (compiled with Erlang/OTP 24)
```

I tried downgrading until 5.2.1 but that did not help either -.- any suggestions are welcome

**Comment by kipcole9** [comment]
That's ... surprising.  I don't have an OpenBSD instance handy but basically what I deduce is that `Money.to_currency/3` is returning an error tuple that happens to match my "standard" error tuple shape but not a format I'm expecting.

While I investigate, it would be helpful if possible for you to change the call from `Money.to_currency!/3` to `Money.to_currency/3` (ie the "non !" version) and see what the return is.

It looks like it probably relates to retrieving external exchange rates but I can't be certain just yet.

Sorry for the inconvenience.

**Comment by kipcole9** [comment]
A couple more things to try:

1. In this [commit](https://github.com/kipcole9/money/commit/c2e30470e14836769c71cc436432f728c260c478) I have changed the return message to ensure that it isn't my usual error tuple shape.  If you don't mind using `{:ex_money, GitHub: "kipcole9/money"} in your `deps` that might make things a little clearer.

2. A connection error should be logged. Can you check if a log message was emitted? If not, perhaps you could configure `ex_money` with appropriate log levels like this:
```elixir
config :ex_money,
  ...
  log_failure: :warn,
  ...
```

**Comment by fbettag** [comment]
so i traced it down that it's working with version 5.0.1. i will try your commit now. one sec.

**Comment by fbettag** [comment]
so with that i am getting

```
** (CaseClauseError) no case clause matching: {:error, "Failed to connect: {:options, {:socket_options, [{:log_level, :alert}, :inet, {:packet_size, 0}, {:packet, 0}, {:header, 0}, {:active, false}, {:mode, 
:binary}]}}"}
```

log_failure is already on :warn.

**Comment by kipcole9** [comment]
I have pushed another [commit](https://github.com/kipcole9/money/commit/71fd78066718aa517c468f16b167b5a73c877df3) that should fix the case clause error and, I think, give a clearer understanding of the failure to connect.

I'm also concerned you're not seeing a log message. Is your log level set to `:warn`? If its set to `:error` or higher then the `ex_money` configuration will need to change to match, at least for now.

The underlying issue here is that there is a failure to connect to the exchange rates server. I don't know why that is, but I hope this update will give us more to debug.

Can you also try `curl https://openexchangerates.org/api/latest.json?app_id=YOUR_APP_ID` on your production machines where `YOUR_APP_ID` is your open exchange rates app_id?

**Comment by kipcole9** [comment]
`mix deps.update ex_money` required to get the latest commit.

**Comment by kipcole9** [comment]
If it is working on `ex_money 5.0.1` as you say then I suspect the difference is that that in [ex_money 5.2](https://github.com/kipcole9/money/blob/v5.7.2/CHANGELOG.md#money-v520) introduced a `:verify_peer` configuration option that default to `true`. This closes a security hole because the default exchange rate retrieve uses `:httpc` which by default does not verify peer on ssl connections.

Therefore if the error persists would you mind:

1. in `iex` on your production machine run `Money.ExchangeRates.Retriever.certificate_store()` and put the results here
2. In `iex` on your production machine run `Money.ExchangeRates.config()` and put the results here
3. Set `verify_peer: false` in your `ex_money` config (`config.exs` or `prod.exs`, however you are doing configuration) and try again.

**Comment by kipcole9** [comment]
One last thought, is it also possible that your `/etc/ssl/cert.pem` file is out of date?  Take a look and see if it contains a certificate described as `=== /O=Digital Signature Trust Co./CN=DST Root CA X3`? That's the root certificate that expired on September 31st this year.

**Comment by fbettag** [comment]
Good morning,

so my cert.pem is up to date, OpenBSD 7.0 was release two months ago, and i can find the correct root CA in the cert.pem:

```
$ grep "Digital Signature Trust Co." /etc/ssl/cert.pem                                                                                                                                                         
### Digital Signature Trust Co.
=== /O=Digital Signature Trust Co./CN=DST Root CA X3
        Subject: O=Digital Signature Trust Co., CN=DST Root CA X3
```

i updated the mix.exs to your github repo again.

```
** (exit) an exception was raised:
    ** (Money.ExchangeRateError) Failed to connect to https://openexchangerates.org/api/latest.json?app_id=5ba81c32515042748556fb2567f2c175: {:options, {:socket_options, [{:log_level, :alert}, :inet, {:packet_si
ze, 0}, {:packet, 0}, {:header, 0}, {:active, false}, {:mode, :binary}]}}
        (ex_money 5.7.2) lib/money.ex:1745: Money.to_currency!/3
```

```
iex(1)> Money.ExchangeRates.Retriever.certificate_store()
'/data/portal/_build/prod/lib/castore/priv/cacerts.pem'
iex(2)> Money.ExchangeRates.config()
%Money.ExchangeRates.Config{
  api_module: Money.ExchangeRates.OpenExchangeRates,
  cache_module: Money.ExchangeRates.Cache.Ets,
  callback_module: Money.ExchangeRates.Callback,
  log_levels: %{failure: :warn, info: :info, success: nil},
  preload_historic_rates: nil,
  retrieve_every: 3600000,
  retriever_options: %{
    app_id: "xxxx",
    url: "https://openexchangerates.org/api"
  },
  verify_peer: true
}
```

Using curl works like a charm, i highly doubt it's something Operating System + TLS related.

```
curl -v https://openexchangerates.org/api/latest.json                                                                                                                                                      
*   Trying 54.164.15.182:443...
* Connected to openexchangerates.org (54.164.15.182) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: CN=openexchangerates.org
*  start date: May  2 00:00:00 2021 GMT
*  expire date: May 31 23:59:59 2022 GMT
*  subjectAltName: host "openexchangerates.org" matched cert's "openexchangerates.org"
*  issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
*  SSL certificate verify ok.
> GET /api/latest.json HTTP/1.1
> Host: openexchangerates.org
> User-Agent: curl/7.79.0
> Accept: */*
```

Setting verify_peer to false seems to work. This is super weird.

**Comment by kipcole9** [comment]
@fbettag, thanks for hanging in there. Given that its working for you in development but not production, and given that you are using `castore`, then I think we know that the issue is ssl verification (evidence: Works when you set `verify_peer: false`)

Is there some possibility that the OTP version on your development and production machines are different?  That at least I can test if you confirm the OTP versions.

I think also I need to make it easy to provide an alternative to using the builtin `:httpc` client as well, but that will take some time.

**Comment by fbettag** [comment]
Yes they are different, check my initial/first message on this issue :)

**Comment by kipcole9** [comment]
Arrrgghhhhhhhh. I was following [this advice](https://erlef.github.io/security-wg/secure_coding_and_deployment_hardening/ssl.html) which suggested that the `:customize_hostname_check` would work on OTP21. Turns out if appears not to work until OTP 22.

I have made some more commits and tested on an Ubuntu server that was showing exactly the same symptoms that you showed.  With the latest updates, it appears to be working now on OTP 21.  Basically I'm using a different host verification path on OTP 21 and earlier.

Would you mind running one more test on your side?

I need to add some documentation to make some notes about this and then if your testing is ok I will publish a new release.

**Comment by fbettag** [comment]
works as expected :) thank you

**Comment by kipcole9** [comment]
Published [ex_money version 5.7.3](https://hex.pm/packages/ex_money/5.7.3) with the following changelog entry:

### Bug Fixes

* Fixes retrieving exchange rates on OTP releases before OTP 22.  Thanks to @fbettag for the report, collaboration and patience.

**Comment by fbettag** [comment]
nah thanks man for providing such a great library and support
